### ISSUE TITLE
Chrome extension memory usage keep growing when frequent use

### ISSUE THREAD

**Issue by cloverzrg** [bitwarden/clients#2249]
### Steps To Reproduce

In chrome, every time click the icon to open the bitwarden extension, the memory usage will increase by about 10M. When the browser is started, the extension's memory usage is only 44M. Click the icon to open the extension and close, repeat it in many, the memory usage will grow to 1G

1. launch chrome
2. open task manager and notice the bitwarden extension memory usage
3. [repeat]click the icon to open and then close bitwarden extension in different tab or in the same tab
4. close all page tab
5. the memory usage still high 

### Expected Result

1.when close all tab, the memory usage should down
2.the memory usage should not increase when repeat open and close the extension in same tab

### Actual Result

1.when close all tab, the memory usage still high
2.the memory usage will increase about 10M each open and close the extension in same tab

### Operating System

macOS

### Web Browser

Chrome

### Browser Version

96.0.4664.110 (x86_64)

### Build Version

1.55.0

**Comment by furt555nv** [comment]
I have same problem. This extension using so much RAM.

**Comment by djsmith85** [comment]
I wanted to report back on this and say that is actively being investigated. The same goes for the issues https://github.com/bitwarden/browser/issues/2198 and https://github.com/bitwarden/browser/issues/1770

**Comment by qwx** [comment]
so i try some debug function and found every time open the window, the bitwarden try to load 10M+ data.
I don't have a lot of passwords save in my vault, so i think the data may related to weak password check.
I really dont know about gc detail in chrome, but the timeline shows that the js heap is growing about 14M everytime when i click the icon,but gc only collect 4M memory. other 10M memory never destory after the extension windows close. Even i force chrome run gc, the data still in memory.
in the heap view i guess this is weak password
![image](https://user-images.githubusercontent.com/6256707/154187448-41c5fc6c-5e5e-4eb3-a32d-9012a89d0816.png)
in the timeline
![image](https://user-images.githubusercontent.com/6256707/154187501-fc7a84a5-58d5-4459-99aa-b68a432fcd0c.png)
ps. i think nodes is growing too , not only the js heap.
@djsmith85 perhaps this would help, sorry about my english writing.

**Comment by samatcolumn** [comment]
I am seeing the Bitwarden Chrome extension take _at least_ 450MB at all times and often up to 950MB of memory unless I forcibly kill it. IMO both of those are too high, and it's really putting a strain on my 8GB RAM MacBook.

**Comment by kspearrin** [comment]
Seems related to and https://github.com/bitwarden/browser/issues/1770

I am able to reproduce the issue on Windows, however, I am also able to reproduce the issue with other extensions that I also have installed that also have a popup window like ours (I tested with uBlock origin). Can you test this? I am not sure what level of memory management capability we have here.

**Comment by pratikabu** [comment]
I've same problem and was about to raise a ticket. I just killed bitwarden as it was using 1 gb memory.

**Comment by xbec65** [comment]
Same, It consumes 1-2 Gb memory.

**Comment by Muvuk** [comment]
I can see that there is a new version of the extension, but the issue still occurs.

**Comment by ruzrobert** [comment]
Latest Chrome, getting 2.5 GB of memory usage 😬 
![image](https://user-images.githubusercontent.com/25375065/167290791-65c04b2c-6e63-46f4-af23-5da1d5fb1dcd.png)

**Comment by ichramm** [comment]
Can confirm it tops 1 GB in Brave.

For those who don't want to restart the browser, disabling/enabling the extension will reset it (you will have to input your master key tough).

**Comment by cloverzrg** [comment]
> 



> Can confirm it tops 1 GB in Brave.
> 
> For those who don't want to restart the browser, disabling/enabling the extension will reset it (you will have to input your master key tough).

you can just kill the ext

**Comment by pc00per** [comment]
Can confirm the same on Chrome/Chromium. The initial memory usage itself is in the range of 150 to 250 megs. I don't think this is a sane behaviour of a browser extension at all.
When I sort memory footprint in descending order, BitWarden extension has always been the first place.

**Comment by qwx** [comment]
So after reading code, I found that there maybe a memory leak in popup.
In popup/app.component.ts ,function ngOnInit
The code 
```
    this.stateService.activeAccount.subscribe((userId) => {
      this.activeUserId = userId;
    });
```
Add an observers to activeAccount (BehaviorSubject) and never unsubscribe.
After click the icon so many times, the observers is increasing and never fall back.
So everytime popup loaded, all memory in popup will never been freed,incloud weak password and user data.
Unfortunately, popup using zxcvbn to detect weak password, so every memory leak will use 10MB memory at least. 
I dont kown it is correct or not,so please help figure this out.

@kspearrin really need your help.
Sorry about my english writing.

**Comment by pc00per** [comment]
@qwx [do you mean this?](https://github.com/bitwarden/desktop/blob/ef60112855237c7fc268abed1a15f067ff9128b4/src/app/app.component.ts#L129-L132)

**Comment by qwx** [comment]
> @qwx [do you mean this?](https://github.com/bitwarden/desktop/blob/ef60112855237c7fc268abed1a15f067ff9128b4/src/app/app.component.ts#L129-L132)

yes.

**Comment by qwx** [comment]
After trying some localhost debug, if i remove subscribe and MatchMedia listeners, memory stop growing.
But It's a huge problem for me to fix the bug and make a pull request. I dont know how to fix the problem in the right way. I know the code is useful :(
So i hope someone can do some research and fix the bug I found, and if the memory leak is still there, I will try another way to 
find the memory leak point.

**Comment by pc00per** [comment]
> After trying some localhost debug, if i remove subscribe and MatchMedia listeners, memory stop growing. But It's a huge problem for me to fix the bug and make a pull request. I dont know how to fix the problem in the right way. I know the code is useful :( So i hope someone can do some research and fix the bug I found, and if the memory leak is still there, I will try another way to find the memory leak point.

Just provide your workaround here and devs will take care of the rest.

**Comment by qwx** [comment]
> > After trying some localhost debug, if i remove subscribe and MatchMedia listeners, memory stop growing. But It's a huge problem for me to fix the bug and make a pull request. I dont know how to fix the problem in the right way. I know the code is useful :( So i hope someone can do some research and fix the bug I found, and if the memory leak is still there, I will try another way to find the memory leak point.
> 
> Just provide your workaround here and devs will take care of the rest.

Already provided previously~

**Comment by addisonbeck** [comment]
Thank you for all the input & patience folks. Proper subscription management is definitely an issue here, and definitely creating a memory leak. I'm going to be taking a look at the subscriptions across clients next week (including the work already done by @qwx, thank you) and getting those handled properly. We aren't completely sure subscription management is the only issue (this issue predates the code examples seen so far, especially StateService), but there is more work planned in this area that I will elaborate on afterwards if the subscription cleanup doesn't alleviate everything.

**Comment by qwx** [comment]
> Thank you for all the input & patience folks. Proper subscription management is definitely an issue here, and definitely creating a memory leak. I'm going to be taking a look at the subscriptions across clients next week (including the work already done by @qwx, thank you) and getting those handled properly. We aren't completely sure subscription management is the only issue (this issue predates the code examples seen so far, especially StateService), but there is more work planned in this area that I will elaborate on afterwards if the subscription cleanup doesn't alleviate everything.

MatchMedia listener also lead memory leak, https://github.com/bitwarden/clients/blob/master/apps/browser/src/services/browserPlatformUtils.service.ts#L363-L367
But i think this is easy to fix, thank you so much！

**Comment by samatcolumn** [comment]
@addisonbeck thanks for validating our concerns, I appreciate your positive attitude about maintaining this extension!

**Comment by qwx** [comment]
> @qwx if you're still available with a local environment I'd be interested to know how the branch above works on your machine. The media match leak should be stopped, and I'm not seeing memory growing on my own machine anymore.
> 
> I don't think the `stateService` subscription is relevant to this. It's not manually unsubscribed from, but `AppComponent` is destroyed each time the popup is closed (or at least it should be) so that subscription should be destroyed along with it. This is where we'll look next, regardless.


I dont have local environment but I use dev tools and edit file in plugin file folder. But I will try to do some test on your branch, thank you very much.

**Comment by addisonbeck** [comment]
The patch for the event listener and the patch to unsubscribe from `StateService.activeAccount` in `AppComponent` will be included in our release in August. Hopefully those two fixes gets most existing issues resolved. I'll check back then and see how folks on this thread are feeling. Thanks again for all the patience and feedback!

**Comment by jasonkylejacobs** [comment]
Updated my Chrome extension to version 2022.8.0, but the issue still seems to be there. Everytime I open the extension, the memory usage increases, and doesn't get released at any point.

**Comment by qwx** [comment]
yes, I did some test, the code in apps/browser/src/popup/app.component.ts function ngOnDestroy not working, maybe the code in ngOnDestroy never running?
Or maybe we can unsubscribe all old observers  when popup show?
---
The Theming service works perfect, event listener doesn't create memory leak now.
Thanks for your great work! @addisonbeck 

---
after walking through this issue https://github.com/bitwarden/clients/issues/3166
I wondering if there is a memory leak,all passwords will left in memory all the time until you close your browser?
@djsmith85 maybe this information can help, thank you a lot.

**Comment by collinlove** [comment]
> Updated my Chrome extension to version 2022.8.0, but the issue still seems to be there. Everytime I open the extension, the memory usage increases, and doesn't get released at any point.

I'm using the Safari extension on the latest version of Monterey. Just updated to 8.1 and this issue is still there. :(

**Comment by fuomag9** [comment]
Referencing my other response, this is still not fixed in safari
https://github.com/bitwarden/clients/issues/1770#issuecomment-1217783178

**Comment by crazyserb** [comment]
Any update on this?

I'm on v2022.9.1 and am still experiencing this on Chrome.

**Comment by yesitisme** [comment]
is there any progress on this?
I noticed this issue on Firefox and on chrome but Firefox is much worse then chrome.

the interesting thing is that I only have this issue on one device multiple browsers. 
I use it on many devices without an issue
![2024-08-29 12_37_51-Task Manager](https://github.com/user-attachments/assets/9069e1f4-2d48-45df-a996-c305650dfed8)

**Comment by shawngmc** [comment]
I'm also not convinced this issue is actually fixed. While I do have a lot of tabs (100+ at current count) and the browser has been open for a little while, 4GB seems... excessive.

![image](https://github.com/user-attachments/assets/8653ccb7-60a4-4659-906b-332985244e27)

I tried doing an Inspect > Memory > Take snaphot, but then in trying to load the snapshot, either the background worker process crashed or the Bitwarden window timed out and closed, taking the browser tools window with it.

**Comment by KlutzyBubbles** [comment]
Is there any reason this was closed? I can't immediately find anything linked that suggests it was fixed and i too am still having the issue of ram usage just going up and up sometimes up to 2-3GB.

**Comment by cloverzrg** [comment]
> Is there any reason this was closed? I can't immediately find anything linked that suggests it was fixed and i too am still having the issue of ram usage just going up and up sometimes up to 2-3GB.

This issue has not been discussed for a long time, it may not be the same problem, please post a new issue

**Comment by KlutzyBubbles** [comment]
> > Is there any reason this was closed? I can't immediately find anything linked that suggests it was fixed and i too am still having the issue of ram usage just going up and up sometimes up to 2-3GB.
> 
> This issue has not been discussed for a long time, it may not be the same problem, please post a new issue, and describe the 'Steps To Reproduce'. if is a same problem, i can reopen the issue

My steps to reproduce are very similar to this issue hence why i commented instead of opening a new issue.

'[repeat]click the icon to open and then close bitwarden extension in different tab or in the same tab'

Still increases ram usage with no end in sight, until i restart chrome or kill the extension.

If you think it still warrants a new issue i will do, however i am still a bit confused as to where exactly this issue was solved other than being old (which seems to only indicate that it hasn't been addressed and is still an issue).

**Comment by vmirage** [comment]
Locking the vault seems to shrink the memory usage

**Comment by trmartin4** [comment]
Hello,

With extension versions `2024.12.0` and higher, we have introduced a change that should reduce the incremental memory footprint of the Chrome extension.  We would like to get feedback on whether this extension version reduces the memory footprint for those reporting the issue here.

**Comment by qwx** [comment]
> Hello,
> 
> With extension versions `2024.12.0` and higher, we have introduced a change that should reduce the incremental memory footprint of the Chrome extension. We would like to get feedback on whether this extension version reduces the memory footprint for those reporting the issue here.

I think problem has been solved, memory size is not growing after repeat open extension in chrome.
can you provide which pr solve this problem? Thanks a lot.

**Comment by vmirage** [comment]
Using 2024.12.2 and problem doesn't seem to be fixed. Bitwarden extension's memory usage still growing after using chrome for few days without closing.

**Comment by trmartin4** [comment]
@qwx the PR that we believe addressed this was https://github.com/bitwarden/clients/pull/11585.

@vmirage to make sure we are addressing the same problem, is there any particular action you're taking that you've been able to correlate with the memory increase, or is it a gradual increase over time?

**Comment by vmirage** [comment]
I have no idea what correlates with the memory increase but one thing for sure is that if I don't close chrome for days, Bitwarden extension memory usage will keep growing. A fresh start, it's about 100MB. I've seen it grow till over 1GB. I normally either terminate the process or just lock the vault to solve the problem for now.

**Comment by Helderjfl** [comment]
> I have no idea what correlates with the memory increase but one thing for sure is that if I don't close chrome for days, Bitwarden extension memory usage will keep growing. A fresh start, it's about 100MB. I've seen it grow till over 1GB. I normally either terminate the process or just lock the vault to solve the problem for now.

Same problem. There's some serious memory leak/management problem on this extension.

**Comment by trmartin4** [comment]
Hello!  For those still experiencing this, can you confirm that the memory usage actually grows indefinitely?  What we have observed internally is that there is a fairly large baseline memory usage (a lot of which appears to be introduced by Chrome itself), and it does increase when being used.  However, in our observation, it reaches a point at which Chrome decides to reclaim memory and the usage stabilizes.  In other words, memory usage is not infinite, but rather it exists in a band that has a lower and upper limit.

**Comment by JMVRy** [comment]
I will mention that I had the exact same problem, but on the Firefox extension instead. I noticed that Firefox was taking up 4GB in total, so I looked at the PIDs to see what was causing it. The `about:processes` list said that the specific PID was "extensions" and `about:memory` said that Bitwarden specifically was taking up 2.5-3.5GB. The memory being used increased when locking as well. My solution was using `about:processes` to kill extensions, and after that it now only uses 30mb, which is so much better.

Even after spamming the extension to make the popup open many times, it hasn't increased.

so basically, just kill the extension process and it'll work, ez pz

**Comment by trmartin4** [comment]
Hello.  We are going to close this issue as this does not appear to be an active issue right now.  Please feel free to re-open or create another issue if you experience this issue again.
